### PR TITLE
Generate stream metadata from releases

### PR DESCRIPTION
**Pull request by sinnykumari**

### CODE REVIEW

File: main.go
Comment (lucab) [inline main.go]: `os.Exit` and `defer Close` don't play well together.

My suggestion is to carve the logic into a `func run() error {}` and let `main` only perform the top-level call and a single `os.Exit`.

File: release.go
Comment (arithx) [inline release.go]: Might be worthwhile to use a [map of strings to a generic architecture type](https://github.com/coreos/mantle/blob/master/cmd/plume/types.go#L104-L106) over a wrapper struct.

File: main.go
Comment (arithx) [inline main.go:235]: Is it intended that `stream.json` is still produced if `overrideFilename` is specified? I would expect that if specified the output would be routed to that location instead of `stream.json`.

Comment (sinnykumari) [reply main.go:235]: As far as I understand yes, stream.json will be automatically generated in both cases - a) when release.json is updated  b) when an override file is specified.
As per https://github.com/coreos/fedora-coreos-tracker/issues/192#issue-450551098 , stream.json is what user will be referring to download latest media . Override file be a partial file(following stream,json schema) which will be used only when we want users to download some other version than what is available in release.json .Some detail on override file is in https://github.com/coreos/fedora-coreos-tracker/issues/98 . @bgilbert did I get things right?

**Comment by sinnykumari** [comment]
Thanks @lucab and @arithx for review! Addressed review comment in https://github.com/sinnykumari/fedora-coreos-stream-generator/pull/1/commits/2899dae07db40fcc952b197da1d7d3b31c6a02da

File: main.go
Comment (arithx) [reply main.go:235]: Ah, I misread this initially as an additional output file. Don't mind me.

File: main.go
Comment (arithx) [inline main.go]: The golang style of formatting makes this block a bit hard to read at a glance. Might read better doing something like so:
```golang
if az := releaseArch.Media.Azure.Images; az != nil && az.Global != nil && az.Global.Image != nil {
    ...
}
```

File: release_index.go
Comment (lucab) [inline release_index.go:56]: It would be better to have a check that `len(relIndex.Releases) > 0`, this may panic.

File: main.go
Comment (lucab) [inline main.go]: Minor style suggestion: you can write this as `if err := run(); err != nil {...}`.

**Comment by lucab** [comment]
Just two minor things, looks good otherwise. I didn't spot any logic mistake in here, but I didn't look too hard...

File: main.go
Comment (sinnykumari) [reply main.go]: Good idea, fixed it!

File: release_index.go
Comment (sinnykumari) [reply release_index.go:56]: Indeed, nice catch. fixed!

File: main.go
Comment (sinnykumari) [reply main.go]: I am still picking up Go, thanks for the suggestion, fixed!

File: main.go
Comment (jlebon) [inline main.go]: Should this be updated for the new name? E.g. `--releases` ?

File: main.go
Comment (jlebon) [inline main.go:187]: Does this transparently support local paths for testing?

File: main.go
Comment (jlebon) [inline main.go]: Minor: could move this down to where we actually decode below?

File: main.go
Comment (jlebon) [inline main.go]: Minor: should probably make this a switch, and otherwise default to output to the console or something.

File: main.go
Comment (jlebon) [inline main.go]: Minor: `s/Open/open/`

**Review by jlebon — COMMENTED** [review]
Just some minor comments from a quick scan!

**Comment by arithx** [comment]
I tested this against the meta translation tool I'm writing and it worked. I would suggest a patch dropping the `Artifacts` type and using a `map[string]*ImageFormat` where you're currently declaring them. That way the `Artifacts` type doesn't have to specify every single format for them to be picked up.

Here's an example of the patch I hacked together:

```
From 84c61de26b553b775bff1eebb358e95129bbf52c Mon Sep 17 00:00:00 2001
From: Stephen Lowrie
Date: Fri, 5 Jul 2019 10:27:15 -0500
Subject: [PATCH] *: switch artifacts to map

This allows the tool to pick up different artifact formats without
needing to add additional keys.
---
 common.go  | 14 --------------
 release.go | 16 ++++++++--------
 stream.go  |  4 ++--
 3 files changed, 10 insertions(+), 24 deletions(-)

diff --git a/common.go b/common.go
index 310266d..6af7481 100644
--- a/common.go
+++ b/common.go
@@ -5,20 +5,6 @@ type Metadata struct {
 	LastModified string `json:"last-modified"`
 }
 
-// Artifacts contains various artifacts
-type Artifacts struct {
-	Raw     *ImageFormat `json:"raw.xz,omitempty"`
-	Qemu    *ImageFormat `json:"qcow.xz,omitempty"`
-	Ova     *ImageFormat `json:"ova,omitempty"`
-	Tar     *ImageFormat `json:"tar.gz,omitempty"`
-	Vmdk    *ImageFormat `json:"vmdk.xz,omitempty"`
-	Vdi     *ImageFormat `json:"vdi.xz,omitempty"`
-	Iso     *ImageFormat `json:"iso,omitempty"`
-	Pxe     *ImageFormat `json:"pxe,omitempty"`
-	InstIso *ImageFormat `json:"installer.iso,omitempty"`
-	InstPxe *ImageFormat `json:"installer-pxe,omitempty"`
-}
-
 // ImageFormat contains Disk image details
 type ImageFormat struct {
 	Disk      *ImageType `json:"disk,omitempty"`
diff --git a/release.go b/release.go
index 6e14ac8..56d3ffe 100644
--- a/release.go
+++ b/release.go
@@ -29,20 +29,20 @@ type ReleaseMedia struct {
 
 // ReleaseAws contains AWS image information
 type ReleaseAws struct {
-	Artifacts Artifacts                      `json:"artifacts"`
+	Artifacts map[string]*ImageFormat        `json:"artifacts"`
 	Images    *map[string]*ReleaseCloudImage `json:"images"`
 }
 
 // ReleaseDigitalOcean DigitalOcean image detail
 type ReleaseDigitalOcean struct {
-	Artifacts Artifacts `json:"artifacts"`
-	Image     string    `json:"image"`
+	Artifacts map[string]*ImageFormat `json:"artifacts"`
+	Image     string                  `json:"image"`
 }
 
 // ReleaseAzure Azure image detail
 type ReleaseAzure struct {
-	Artifacts Artifacts           `json:"artifacts"`
-	Images    *ReleaseAzureImages `json:"images"`
+	Artifacts map[string]*ImageFormat `json:"artifacts"`
+	Images    *ReleaseAzureImages     `json:"images"`
 }
 
 // ReleaseAzureImages Azure image detail
@@ -52,8 +52,8 @@ type ReleaseAzureImages struct {
 
 // ReleaseGcp GCP image detail
 type ReleaseGcp struct {
-	Artifacts Artifacts `json:"artifacts"`
-	Image     *string   `json:"image"`
+	Artifacts map[string]*ImageFormat `json:"artifacts"`
+	Image     *string                 `json:"image"`
 }
 
 // ReleaseCloudImage cloud image information
@@ -63,5 +63,5 @@ type ReleaseCloudImage struct {
 
 // ReleaseTargetPlatform target platforms
 type ReleaseTargetPlatform struct {
-	Artifacts Artifacts `json:"artifacts"`
+	Artifacts map[string]*ImageFormat `json:"artifacts"`
 }
diff --git a/stream.go b/stream.go
index 23fded1..4f6a3cc 100644
--- a/stream.go
+++ b/stream.go
@@ -30,8 +30,8 @@ type StreamArtifacts struct {
 
 // StreamMediaDetails contains image artifact and release detail
 type StreamMediaDetails struct {
-	Release string    `json:"release"`
-	Formats Artifacts `json:"formats"`
+	Release string                  `json:"release"`
+	Formats map[string]*ImageFormat `json:"formats"`
 }
 
 // StreamImages contains images available in cloud providers
-- 
2.21.0
```

File: main.go
Comment (sinnykumari) [reply main.go]: fixed

File: main.go
Comment (sinnykumari) [reply main.go]: yeah, sounds good idea. Printing default to stdout now  https://github.com/sinnykumari/fedora-coreos-stream-generator/pull/1/commits/1da0423afb9e2c16bf9b762e71ee5f9ab244aae3

File: main.go
Comment (sinnykumari) [reply main.go]: fixed

File: main.go
Comment (sinnykumari) [reply main.go:187]: Not yet, fixed it in https://github.com/sinnykumari/fedora-coreos-stream-generator/pull/1/commits/1da0423afb9e2c16bf9b762e71ee5f9ab244aae3

File: main.go
Comment (sinnykumari) [reply main.go]: fixed

**Comment by sinnykumari** [comment]
Thanks @arithx for the patch! it looks good and have pushed same in https://github.com/sinnykumari/fedora-coreos-stream-generator/pull/1/commits/5e56fa07e57355e9cd94d588a16b06936de5932f

**Comment by sinnykumari** [comment]
Can we have a final review on this PR before we merge it to master?

File: main.go
Comment (jlebon) [inline main.go:272]: Minor: could dedupe this logic by having e.g. an `out` variable that you set to either `os.Stdout` or `streamFile`.

**Review by jlebon — COMMENTED** [review]
Just one minor comment, otherwise LGTM!

File: main.go
Comment (sinnykumari) [reply main.go:272]: Thanks, fixed it

**Review by jlebon — APPROVED** [review]
Last comment: in the interest of keeping the git history clean, WDYT about squashing it all to a single "initial commit" patch to seed master with? I find my initial git commits are usually of low quality as I hack things together, and so becomes mostly noise.

Up to you though, it mostly comes down to personal preference. :)

**Comment by sinnykumari** [comment]
I am also in favor if keeping git history clean. Squashed all commits to single commit.

Will wait for some time before merging to address any final review comment

**Comment by sinnykumari** [comment]
merged, thanks all for review!

**Pull request merged**
